smm/lotus_adapter.go

```diff
@@ -0,0 +1,266 @@
+
+        case <-adapter.context.Done():
+            return
+        }
+    }
+}
+
+func (adapter lotusAdapter) notifyListeners(changes []*store.HeadChange) {
+    stateChanges, err := stateChangesFromHeadChanges(changes)
+    if err != nil {
+        // TODO: Log, nothing else to do
+        return
+    }
+
+    for _, stateChange := range stateChanges {
+        for listener, _ := range adapter.listeners {
+            listener.OnChainStateChanged(stateChange)
+        }
+    }
+}
+
+func (adapter lotusAdapter) SubscribeMiner(ctx context.Context, cb StorageMiningEvents) error {
+    adapter.listeners[cb] = struct{}{}
+    return nil
+}
+
+func (adapter lotusAdapter) UnsubscribeMiner(ctx context.Context, cb StorageMiningEvents) error {
+    _, found := adapter.listeners[cb]
+    if found {
+        delete(adapter.listeners, cb)
+    }
+    return nil
+}
+
+func (adapter lotusAdapter) callMinerActorMethod(ctx context.Context, method uint64, payload []byte) (cid cid.Cid, err error) {
+    // TODO: Validate that using the miner/worker address for both To and From is allowed
+    msg := types.Message{
+        To:       adapter.address,
+        From:     adapter.address,
+        Method:   method,
+        Params:   payload,
+        // TODO: Add ability to control these 'costs'
+        Value:    types.NewInt(1000), // currently hard-coded late fee in actor, returned if not late
+        GasLimit: types.NewInt(1000000 /* i dont know help */),
+        GasPrice: types.NewInt(1),
+    }
+
+    var msgErr error
+    signedMsg, msgErr := adapter.fullNode.MpoolPushMessage(ctx, &msg)
+
+    if msgErr != nil {
+        err = msgErr
+        return
+    }
+    cid = signedMsg.Cid()
+    return
+}
+
+func (adapter lotusAdapter) MostRecentState(ctx context.Context) (stateKey *StateKey, epoch Epoch, err error) {
+    var ts *types.TipSet
+    ts, err = adapter.fullNode.ChainHead(ctx)
+    if err != nil {
+        return
+    }
+    epoch = Epoch(ts.Height())
+    stateKey, err = tipset2statekey(ts)
+    return
+}
+
+func (adapter lotusAdapter) GetMinerState(ctx context.Context, stateKey *StateKey) (state MinerChainState, err error) {
+    var ts *types.TipSet
+    ts, err = statekey2tipset(ctx, stateKey, adapter.fullNode)
+    if err != nil {
+        return
+    }
+    sectors, apiErr := adapter.fullNode.StateMinerSectors(ctx, adapter.address, ts)
+    if apiErr != nil {
+        err = apiErr
+        return
+    }
+    provingSectors, apiErr := adapter.fullNode.StateMinerProvingSet(ctx, adapter.address, ts)
+    if apiErr != nil {
+        err = apiErr
+        return
+    }
+    state.Sectors = make(map[uint64]api.ChainSectorInfo)
+    for _, sectorInfo := range sectors {
+        state.Sectors[sectorInfo.SectorID] = *sectorInfo
+    }
+    for _, sectorInfo := range provingSectors {
+        state.ProvingSet[sectorInfo.SectorID] = struct{}{}
+    }
+    state.Address = Address(adapter.address.String())
+    // fill in this information
```

Comment:
Is this a TODO for this PR? Please either do it, or I'd suggest holding back this code for a future PR that implements this method fully.